foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/mappings/AggregateCollectionMapping.java

```diff
@@ -2526,31 +2526,19 @@ public void preInsert(WriteObjectQuery query) throws DatabaseException, Optimist
 
     /**
      * INTERNAL:
-     * Returns clone of InsertObjectQuery from the reference descriptor, if it is not set - create it.
+     * Returns a clone of InsertObjectQuery from the ClassDescriptor's DescriptorQueryManager or a new one
      */
     protected InsertObjectQuery getInsertObjectQuery(AbstractSession session, ClassDescriptor desc) {
-        DescriptorQueryManager queryManager = desc.getQueryManager();
-        if (!queryManager.hasInsertQuery()) {
-            synchronized (queryManager) {
-                if (!queryManager.hasInsertQuery()) {
-                    queryManager.setInsertQuery(new InsertObjectQuery());
-                }
-            }
-        }
-        InsertObjectQuery insertQuery = queryManager.getInsertQuery();
-        if (insertQuery.getModifyRow() == null) {
-            AbstractRecord modifyRow = new DatabaseRecord();
-            for (int i = 0; i < getTargetForeignKeyFields().size(); i++) {
-                DatabaseField field = getTargetForeignKeyFields().elementAt(i);
-                modifyRow.put(field, null);
-            }
-            desc.getObjectBuilder().buildTemplateInsertRow(session, modifyRow);
-            getContainerPolicy().addFieldsForMapKey(modifyRow);
-            if(this.listOrderField != null) {
-                modifyRow.put(this.listOrderField, null);
-            }
-            insertQuery.setModifyRow(modifyRow);
+        InsertObjectQuery insertQuery = desc.getQueryManager().getInsertQuery();
+        if (insertQuery == null) {
+            insertQuery = new InsertObjectQuery();
+            insertQuery.setDescriptor(desc);
         }
+
+        // Ensure the query has been prepared.
+        insertQuery.checkPrepare(session, insertQuery.getTranslationRow());
+        insertQuery = (InsertObjectQuery)insertQuery.clone();
```

Comment:
this call to clone should be moved to 'else' branch - no need to clone newly created instance, only the one obtained from query manager

Reply:
I wrote it this way to simplify the logic since we need to call ```insertQuery.checkPrepare``` on both the new and the one set on the QueryManager before clone, yes? This is the pattern the previous code used and I was attempting to mirror it.

Reply:
I'd follow pattern from other places, like in [CommitManager](https://github.com/eclipse-ee4j/eclipselink/blob/master/foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/internal/sessions/CommitManager.java#L213)

Reply:
So you are saying that ```insertQuery.checkPrepare``` doesn't need to be called on the new instance? Again, this is fine, but it is yet another deviation from how the code originally ran. The original failure was due to the descriptor not being set on the new instance followed by ```insertQuery.checkPrepare``` (which expected the descriptor set).
I have no problem making these changes, but my first inclination is not to rewrite everything, it's to preserve as much of the original logic as possible. If you think nixing the ```insertQuery.checkPrepare``` is fine, Ill rerun the tests and cross my fingers.